bigtable/metrics.go

```diff
@@ -0,0 +1,426 @@
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"cloud.google.com/go/bigtable/internal"
+	"github.com/google/uuid"
+	gax "github.com/googleapis/gax-go/v2"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	builtInMetricsMeterName = "bigtable.googleapis.com/internal/client/"
+
+	metricsPrefix         = "bigtable/"
+	locationMDKey         = "x-goog-ext-425905942-bin"
+	serverTimingMDKey     = "server-timing"
+	serverTimingValPrefix = "gfet4t7; dur="
+
+	// Monitored resource labels
+	monitoredResLabelKeyProject  = "project_id"
+	monitoredResLabelKeyInstance = "instance"
+	monitoredResLabelKeyTable    = "table"
+	monitoredResLabelKeyCluster  = "cluster"
+	monitoredResLabelKeyZone     = "zone"
+
+	// Metric labels
+	metricLabelKeyAppProfile         = "app_profile"
+	metricLabelKeyMethod             = "method"
+	metricLabelKeyStatus             = "status"
+	metricLabelKeyStreamingOperation = "streaming"
+	metricLabelKeyClientName         = "client_name"
+	metricLabelKeyClientUID          = "client_uid"
+
+	// Metric names
+	metricNameOperationLatencies = "operation_latencies"
+	metricNameAttemptLatencies   = "attempt_latencies"
+	metricNameServerLatencies    = "server_latencies"
+	metricNameRetryCount         = "retry_count"
+)
+
+var (
+	// duration between two metric exports
+	// Not a constant. Modifiable in tests
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName = fmt.Sprintf("go-bigtable/%v", internal.Version)
+
+	// Arrays cannot be constants in Go
+	bucketBounds = []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0, 16.0, 20.0, 25.0, 30.0, 40.0,
+		50.0, 65.0, 80.0, 100.0, 130.0, 160.0, 200.0, 250.0, 300.0, 400.0, 500.0, 650.0,
+		800.0, 1000.0, 2000.0, 5000.0, 10000.0, 20000.0, 50000.0, 100000.0, 200000.0,
+		400000.0, 800000.0, 1600000.0, 3200000.0}
+
+	// All the built-in metrics have same attributes except 'status' and 'streaming'
+	// These attributes need to be added to only few of the metrics
+	// Maps cannot be constants in Go
+	additionalAttributes = map[string][]string{
+		metricNameOperationLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameAttemptLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameServerLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameRetryCount: {
+			metricLabelKeyStatus,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<>hostname
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return "go-" + uuid.NewString() + "@" + hostname, nil
+	}
+
+	// Not a constant. Modifiable in tests
+	exporterOpts = []option.ClientOption{}
+)
+
+type builtinMetricsTracerFactory struct {
+	builtinEnabled bool
```

Comment:
can this just be enabled? I'm not sure what prefixing it with built does

Reply:
Done